llama.cpp

```diff
@@ -4263,3 +4277,44 @@ const char * llama_print_system_info(void) {
 const std::vector<std::pair<std::string, struct ggml_tensor *>>& llama_internal_get_tensor_map(struct llama_context * ctx) {
     return ctx->model.tensors_by_name;
 }
+
+
+void llama_log_set(llama_log_callback log_callback, void * user_data) {
+    g_state.log_callback = log_callback ? log_callback : llama_log_callback_default;
+    g_state.log_callback_user_data = user_data;
+}
+
+#if defined(_MSC_VER) && !defined(vsnprintf)
+#define vsnprintf _vsnprintf
```

Comment:
From https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l?view=msvc-170

> Beginning with the UCRT in Visual Studio 2015 and Windows 10, vsnprintf is no longer identical to _vsnprintf. The vsnprintf function conforms to the C99 standard; _vnsprintf is kept for backward compatibility with older code. The difference is that if you run out of buffer, vsnprintf null-terminates the end of the buffer and returns the number of characters that would have been required, while _vsnprintf doesn't null-terminate the buffer and returns -1. Also, _vsnprintf() includes one more character in the output because it doesn't null-terminate the buffer.

When using LLAMA_LOG_... with result of more than 128 characters this will actually result in `int len = vsnprintf(buffer, 128, format, args);` returning `-1`. Stumbled upon this on my system where I compile with Visual Studio. This will then result in garbled output:

```
>main -m ..\..\..\..\models\ggml-vocab-llama.gguf
main: build = 1096 (103cfaf)
main: seed  = 1693230616
llama_model_loader: loaded meta data with 13 key-value pairs and 0 tensors from ..\..\..\..\models\ggml-vocab-llama.gguf (versio+SCllama_model_loader: - kv   0:                       general.architecture str
llama_model_loader: - kv   1:                       llama.context_length u32
```

The question is why `vsnprintf` is redefined to use `_vsnprintf`?
When it would just use `vsnprintf` the result is as expected:

```
>main -m ..\..\..\..\models\ggml-vocab-llama.gguf
main: build = 1096 (103cfaf)
main: seed  = 1693230594
llama_model_loader: loaded meta data with 13 key-value pairs and 0 tensors from ..\..\..\..\models\ggml-vocab-llama.gguf (version GGUF V1 (support until nov 2023))
llama_model_loader: - kv   0:                       general.architecture str
llama_model_loader: - kv   1:                       llama.context_length u32
```

Reply:
You are right!
I copied the define from an older codebase, which needed to support older VS versions.
I made a PR with the fix here: #2829
Thanks for tracking this down!